### ISSUE TITLE
Events Validation

### ISSUE THREAD

**Issue by azhard4int** [jitsucom/jitsu#540]
### Problem

While using Jitsu, one thing I'd like to have is validation for the events. For example, we are allowing our users to send attributes such as "first_name", "last_name", "created_at" attributes.

The value for the `"created_at": "2021-09-12"` we are recommending our users to send it as a date string, but in case if some of the customers send it as a UNIX timestamp as Integer, the column structure will be changed in the Click house/Postgres when the first-time event is added to the data storage.

What problem this feature is going to solve. Why this problem is important

For SaaS owners who are building their solution on top of Jitsu, this would allow them to add validations for their events based on their Default Attributes for the schema defined for e.g Postgresql / Clickhouse.

Example script from Customer.io

```
// Send this when a user logs in or you otherwise identify them. 
<script type="text/javascript">
    _cio.identify({
        // Required attributes — you must include at least one of the following
        id: 'YOUR_USER_ID_HERE',   // Unique id for the currently signed in user in your application.
                                   // This value is an email address if your workspace uses email
                                   // as an identifier.
        email: 'user@domain.com',  // Email of the currently signed in user.

        // Strongly recommended attributes
        created_at: 1339438758,   // Timestamp in your system that represents when
                                  // the user first signed up. You'll want to send it
                                  // as seconds since the epoch.

        // Example attributes (you can name attributes what you wish)
        first_name: 'John',       // Add any attributes you'd like to use in the email subject or body.
        last_name: 'Smith',       // First name and last name are shown on people pages.
        plan_name: 'premium'      // To use the example segments, set this to 'free' or 'premium'.
    });
</script>
```

Reference link: https://customer.io/docs/javascript-quick-start/

### Solution

Proposed solution: architecture, implementation details etc

1. Add a validations schema under mapping.YAML where we can define the event fields along with their data type. Later, those validations could be performed on the server-side of Jitsu and return with error code 422 status code.
2. Provide a way to store the non-validated/error-based events to the destination. This could, later on, be used by the SaaS companies who are building their solution on top of Jitsu to show the errors in case if their customers are sending their events but are not validated.

**Comment by vklimontovich** [comment]
Hi @azhard4int!

Thank you for raising the issue, it's definitely an interesting use case. Here's what we recommend to do now, and what we can implement further.

_the column structure will be changed in the Click house/Postgres when the first-time event is added to the data storage_

You can avoid that by [setting the column type expicitely](https://jitsu.com/docs/configuration/schema-and-mappings#configuration):

```yaml
destinations:
  destination_name:
    data_layout:
      mappings:
        keep_unmapped: true # if fields that are not explicitly mapped should be kept or removed
        fields:
        - src: /created_at
           action: cast
           type: timestamp with timezone
           column_type: 'timestamp with timezone'
```

In this case, a) a new column will be created with `timestamp with timezone` regardless of incoming data type b) incoming data type will be cast to `timestamp with timezone` during upgrade/insert.

However, it JSON path isn't known in advance, the mapping above won't work. Also, it's not possible to support both timestamps (as int) and YYYY-MM-DD dates.

**Proposed solution**: we already implemented JavaScript functions for table-naming and webhook body, and going to [implement the same for mapping](https://github.com/jitsucom/jitsu/issues/281). With this feature implemented, instead of YAML you will be able to write a custom JS for mapping. The syntax will look like this:

```javascript
export function transform(event) {
    if (event.created_at) {
        if (typeof event.created_at === 'number') {
            event.created_at = new Date(event.created_at);
        } else if (typeof event.created_at === 'string') {
            event.created_at = new Date(Date.parse(event.created_at));
        } else {
            throw new Error(`Invalid created_at format: ${event.created_at}`);
        }
    }
    const meta = {
        created_at: {
            sqlType: `timestampz`
        }
    }
    return ['events', event, meta]
}
```

```yaml
destinations:
  destination_name:
    transform: ./transform.js
```

The other issue is error handling. Both at the moment and with new JS transformations error handling will remain the same. 
 * In first example error will happen if the incoming value won't be casted to `timestamp with timezone` successfully (e.g. integer value instead of 'YYYY-MM-DD' date)
 * Once JS functions will be implemented, an error will be triggered by `throw new Error(...)`

What happens with erroneous events is described in [Jitsu Server Architecture](https://jitsu.com/docs/internals/jitsu-server) article. It might be a little bit outdated, but the folder structure is up-to-date. Faulty events should be written to `events/failed`

**Comment by vklimontovich** [comment]
Will be solved by #282. Keeping this one open so far

**Comment by vklimontovich** [comment]
Closing this one since #282 is closed already
